tests/validators.py

```diff
@@ -0,0 +1,85 @@
+from unittest import TestCase
+
+from adapters.validators import (
+    EMail, Length, Schema, Type, UnicodeNormalize, ValidationError,
+)
+
+
+class ValidatorsTest(TestCase):
+    def setUp(self):
+        class UsernameSchema(Schema):
+            inputs = {'username'}
+            outputs = {'username', 'username_normalized'}
+            validators = [
+                Type('username', str),
+                Length('username', min=4, max=100),
+                UnicodeNormalize('username', out='username_normalized'),
+            ]
+
+        class EMailSchema(Schema):
+            inputs = {'email'}
+            validators = [
+                EMail('email'),
+            ]
+
+        class RegistrationSchema(Schema):
+            depends = [UsernameSchema, EMailSchema]
+
+            def username_in_email(self):
+                if self.username_normalized not in self.email:
+                    raise ValidationError("E-mail must contain username", field='email')
+
+            validators = [
+                username_in_email,
+            ]
+
+        self.UsernameSchema = UsernameSchema
+        self.EMailSchema = EMailSchema
+        self.RegistrationSchema = RegistrationSchema
+
+    def test_simple_validation(self):
+        schema = self.UsernameSchema()
+        validator = schema.validate({'username': 'foobar'})
+        self.assertTrue(validator.is_valid)
+        validator = schema.validate({'username': 'x'})
+        self.assertFalse(validator.is_valid)
+
+    def test_missing_field(self):
+        schema = self.EMailSchema()
+        validator = schema.validate({})
+        self.assertFalse(validator.is_valid)
+
+    def test_partial_validation(self):
+        schema = self.RegistrationSchema()
+        validator = schema.validate({'username': 'foobar', 'email': 'not an email address'})
+        self.assertFalse(validator.is_valid)
+        self.assertTrue(validator['username'].is_valid)
+        self.assertFalse(validator['email'].is_valid)
```

Comment:
The main potential advantage is that it groups all the validity information together instead of having it spread over each individual field. Whether this is actually the more useful way round I'm unsure.